fvm/runtime/reusable_cadence_runtime.go

```diff
@@ -189,11 +196,16 @@ func newReusableCadenceRuntimePool(
 		pool = make(chan *ReusableCadenceRuntime, poolSize)
 	}
 
-	return ReusableCadenceRuntimePool{
+	runtimePool := ReusableCadenceRuntimePool{
 		pool:             pool,
 		config:           config,
 		newCustomRuntime: newCustomRuntime,
 	}
+	if config.CoverageReportingEnabled {
+		runtimePool.CoverageReport = runtime.NewCoverageReport()
```

Comment:
I think the plan/idea is to record coverage across multiple transactions/scripts in the Emulator.

Currently, the Cadence runtime has a configuration option that determines if coverage reporting is enabled in general, for all transactions/scripts. Then, each context that is passed for a transaction/script execution can pass a different coverage report (or e.g. none), that is used if the reporting config option is enabled. This allows each transaction/script execution to have a different coverage report.

Agreed that the runtime configuration option could be removed, and the existing use of it could be replaced with a nil check of the context coverage report, i.e. coverage reporting is considered enabled if a coverage report is passed.

However, even though it does not matter for this use case, I don't thing moving the coverage report from the context (per tx/script) to the config (all tx/scripts) is a good idea, as that will make it harder to have a different coverage report for each tx/script execution.

Regarding the creation of a coverage report for a pool here: My guess is that the idea here was to achieve the goal of recording coverage across multiple transactions/scripts by creating a coverage report for the whole runtime pool, so that the report is used for all runtimes, and thus all transaction/script executions. I guess in the Emulator the coverage report can then be accessed at any time and inspected (the field is exported).

@pattyshack Please feel free to recommend an alternative.

Reply:
> Regarding the creation of a coverage report for a pool here: My guess is that the idea here was to achieve the goal of recording coverage across multiple transactions/scripts by creating a coverage report for the whole runtime pool, so that the report is used for all runtimes, and thus all transaction/script executions. I guess in the Emulator the coverage report can then be accessed at any time and inspected (the field is exported).

@turbolent Your guess is on point! Whether the emulator is used from another library or as a standalone program, we want the developers/users to have access to a shared coverage report, through which they can view code coverage insights. It is necessary that we re-use the same coverage report, throughout the lifecycle of an emulator instance. (simulating scenarios with scripts & transactions or even running integration tests with another library).

For reference, following is the line of code accessing the shared coverage report, in the Emulator:

```go
blockchain.coverageReport = ctx.ReusableCadenceRuntimePool.CoverageReport
```

I hope this clarifies things.

Reply:
Is this only used by the emulator?  if so, maybe just provide a custom runtime implementation via NewCustomReusableCadenceRuntimePool.  All the code will reside within the emulator code base and no change is needed here.

e.g.,
```
type CoverageReportedRuntime struct {
  report *runtime.CoverageReport
  rt runtime.Runtime
}

func (rt *CoverageReportedRuntime) ReadStored(
    address common.Address,
    path cadence.Path,
    context runtime.Context,
) (
    cadence.Value,
    error,
) {
    context.CoverageReport = rt.report
    return rt.rt.ReadStored(address, path, context)
}
```

Reply:
this comment is aligned with what janez is suggesting, in particular, the coverage report will be injected from outside the runtime pool.

A single fvm.Run execution use multiple (two) reusable cadence runtime objects, even when executed sequentially.  You will need to do some sort of state merging if the runtime don't share the same report object.

Reply:
> Is this only used by the emulator?

That is correct. My first approach was to piggy-back on the current implementation and simply provide the existing config flag, e.g:

```go
fvmOptions := []fvm.Option{
	...
	fvm.WithReusableCadenceRuntimePool(
		reusableRuntime.NewReusableCadenceRuntimePool(
			1,
			runtime.Config{
				Debugger:                 blockchain.debugger,
				AccountLinkingEnabled:    true,
                                // Only this line was added in the Emulator
				CoverageReportingEnabled: conf.CoverageReportingEnabled,
			}),
	),
}

ctx := fvm.NewContext(
	fvmOptions...,
)
```

Since a `ReusableCadenceRuntimePool` is already being utilized (I think it was introduced for the debugger, but I am not sure), I merely added the respective config flag to enable coverage reporting. I will try to use your idea of a `NewCustomReusableCadenceRuntimePool`, while accommodating for the debugger at the same time. Thanks for the suggestion :pray: 

> A single fvm.Run execution use multiple (two) reusable cadence runtime objects, even when executed sequentially. You will need to do some sort of state merging if the runtime don't share the same report object.

I don't really get this part :sweat_smile: Is this the case even with your suggestion above?

Reply:
Currently, each fvm.Run execution borrows two cadence runtime objects.  you want to make sure that you create new runtime objects using the same same coverage report / debugger object.

Reply:
Ah, I see. Appreciate the explanation :+1:

Reply:
> if so, maybe just provide a custom runtime implementation via NewCustomReusableCadenceRuntimePool.

if It didn't swallow runtime.config

Reply:
> if It didn't swallow runtime.config

@bluesign I noticed that problem also. However, I think it must have been fixed in master, by https://github.com/onflow/flow-go/commit/5f2bbe2d5c03cd98c11d173fc155d0664debf14a